From bede7e1824a2a0618e5dad6d2a2681504347577a Mon Sep 17 00:00:00 2001
From: Arun Kumar <arunkumar.ev@adlinktech.com>
Date: Fri, 8 May 2020 17:56:16 +0530
Subject: [PATCH 11/11] IMX8QM: Fork separate dts for DVI

Fork HDMI dts file for DVI. Disable audio and add mode_dvi
property to check when to enable HDMI_TX_MODE_DVI

Signed-off-by: Arun Kumar <arunkumar.ev@adlinktech.com>
---
 arch/arm64/boot/dts/adlink/Makefile           |  3 +-
 .../boot/dts/adlink/adlink-lec-imx8qm-dvi.dts | 68 +++++++++++++++++++
 .../dts/adlink/adlink-lec-imx8qm-hdmi.dts     |  2 -
 drivers/gpu/drm/imx/hdp/imx-hdmi.c            |  5 +-
 drivers/gpu/drm/imx/hdp/imx-hdp.c             |  2 +
 drivers/gpu/drm/imx/hdp/imx-hdp.h             |  2 +
 6 files changed, 77 insertions(+), 5 deletions(-)
 create mode 100644 arch/arm64/boot/dts/adlink/adlink-lec-imx8qm-dvi.dts

diff --git a/arch/arm64/boot/dts/adlink/Makefile b/arch/arm64/boot/dts/adlink/Makefile
index 4aa9fc65a5bd..0e518c71f3f8 100644
--- a/arch/arm64/boot/dts/adlink/Makefile
+++ b/arch/arm64/boot/dts/adlink/Makefile
@@ -1,6 +1,7 @@
 dtb-$(CONFIG_ARCH_LEC_IMX8QM) += adlink-lec-imx8qm.dtb \
 				adlink-lec-imx8qm-hdmi.dtb \
-				adlink-lec-imx8qm-ov5640.dtb
+				adlink-lec-imx8qm-ov5640.dtb \
+				adlink-lec-imx8qm-dvi.dtb \
 
 always         := $(dtb-y)
 subdir-y       := $(dts-dirs)
diff --git a/arch/arm64/boot/dts/adlink/adlink-lec-imx8qm-dvi.dts b/arch/arm64/boot/dts/adlink/adlink-lec-imx8qm-dvi.dts
new file mode 100644
index 000000000000..398e2ba39b06
--- /dev/null
+++ b/arch/arm64/boot/dts/adlink/adlink-lec-imx8qm-dvi.dts
@@ -0,0 +1,68 @@
+/*
+ * DVI mode enable, disable ldb display, disable HDMI audio.
+ */
+
+#include "adlink-lec-imx8qm.dts"
+
+/ {
+	sound-hdmi-tx {
+		status = "disabled";
+	};
+
+	sound-amix-sai {
+		status = "disabled";
+	};
+
+	sound-hdmi-arc {
+		status = "disabled";
+	};
+};
+
+&ldb1_phy {
+	status = "disabled";
+};
+
+&ldb1 {
+	status = "disabled";
+};
+
+&i2c1_lvds0 {
+	status = "disabled";
+};
+
+&irqsteer_hdmi {
+	status = "okay";
+};
+
+&hdmi {
+	compatible = "fsl,imx8qm-hdmi";
+	assigned-clocks = <&clk IMX8QM_HDMI_PXL_SEL>,
+					<&clk IMX8QM_HDMI_PXL_LINK_SEL>,
+					<&clk IMX8QM_HDMI_PXL_MUX_SEL>;
+	assigned-clock-parents = <&clk IMX8QM_HDMI_AV_PLL_CLK>,
+							<&clk IMX8QM_HDMI_AV_PLL_CLK>,
+							<&clk IMX8QM_HDMI_AV_PLL_CLK>;
+	fsl,cec;
+	mode-dvi;
+	status = "okay";
+};
+
+&sai_hdmi_tx {
+	assigned-clocks =<&clk IMX8QM_ACM_HDMI_TX_SAI0_MCLK_SEL>,
+			<&clk IMX8QM_AUD_PLL0_DIV>,
+			<&clk IMX8QM_AUD_ACM_AUD_PLL_CLK0_DIV>,
+			<&clk IMX8QM_AUD_ACM_AUD_REC_CLK0_DIV>,
+			<&clk IMX8QM_AUD_SAI_HDMITX0_MCLK>;
+	assigned-clock-parents = <&clk IMX8QM_AUD_ACM_AUD_PLL_CLK0_CLK>;
+	assigned-clock-rates = <0>, <786432000>, <49152000>, <12288000>, <49152000>;
+	fsl,sai-asynchronous;
+	status = "okay";
+};
+
+&spdif1 {
+	assigned-clocks =<&clk IMX8QM_AUD_PLL0_DIV>,
+			<&clk IMX8QM_AUD_ACM_AUD_PLL_CLK0_DIV>,
+			<&clk IMX8QM_AUD_ACM_AUD_REC_CLK0_DIV>;
+	assigned-clock-rates = <786432000>, <49152000>, <12288000>;
+	status = "okay";
+};
diff --git a/arch/arm64/boot/dts/adlink/adlink-lec-imx8qm-hdmi.dts b/arch/arm64/boot/dts/adlink/adlink-lec-imx8qm-hdmi.dts
index d2fc5f0454db..40b56a1b884d 100644
--- a/arch/arm64/boot/dts/adlink/adlink-lec-imx8qm-hdmi.dts
+++ b/arch/arm64/boot/dts/adlink/adlink-lec-imx8qm-hdmi.dts
@@ -24,7 +24,6 @@
 		audio-cpu = <&sai_hdmi_tx>;
 		protocol = <1>;
 		hdmi-out;
-		status = "disabled";
 	};
 
 	sound-amix-sai {
@@ -37,7 +36,6 @@
 		spdif-controller = <&spdif1>;
 		spdif-in;
 		spdif-out;
-		status = "disabled";
 	};
 };
 
diff --git a/drivers/gpu/drm/imx/hdp/imx-hdmi.c b/drivers/gpu/drm/imx/hdp/imx-hdmi.c
index 1dd02d611fc7..ad5acc44c1f2 100644
--- a/drivers/gpu/drm/imx/hdp/imx-hdmi.c
+++ b/drivers/gpu/drm/imx/hdp/imx-hdmi.c
@@ -221,9 +221,10 @@ static int hdmi_scdc_tmds_config(struct imx_hdp *hdp)
 		/* Default work in HDMI1.4 */
 		buff = 0;
 		hdp->hdmi_type = HDMI_TX_MODE_HDMI_1_4;
-	 }
+	}
 
-	hdp->hdmi_type = HDMI_TX_MODE_DVI;
+	if(hdp->mode_dvi)
+		hdp->hdmi_type = HDMI_TX_MODE_DVI;
 
 	data_in.buff = &buff;
 	data_in.len = 1;
diff --git a/drivers/gpu/drm/imx/hdp/imx-hdp.c b/drivers/gpu/drm/imx/hdp/imx-hdp.c
index fbba47131e09..267b5cb38fc7 100644
--- a/drivers/gpu/drm/imx/hdp/imx-hdp.c
+++ b/drivers/gpu/drm/imx/hdp/imx-hdp.c
@@ -1505,6 +1505,8 @@ static int imx_hdp_imx_bind(struct device *dev, struct device *master,
 	if (IS_ERR(hdp->mem.rst_base))
 		dev_warn(dev, "Failed to get HDP RESET base register\n");
 
+	hdp->mode_dvi = of_property_read_bool(pdev->dev.of_node, "mode-dvi");
+
 	hdp->is_cec = of_property_read_bool(pdev->dev.of_node, "fsl,cec");
 
 	hdp->is_digpll_dp_pclock =
diff --git a/drivers/gpu/drm/imx/hdp/imx-hdp.h b/drivers/gpu/drm/imx/hdp/imx-hdp.h
index 52678a0bc0e3..0dc8ab6d46d4 100644
--- a/drivers/gpu/drm/imx/hdp/imx-hdp.h
+++ b/drivers/gpu/drm/imx/hdp/imx-hdp.h
@@ -261,6 +261,8 @@ struct imx_hdp {
 	VIC_PXL_ENCODING_FORMAT format;
 	bool hdr_metadata_present;
 	bool hdr_mode;
+
+	bool mode_dvi;
 };
 
 void imx_hdp_register_audio_driver(struct device *dev);
-- 
2.17.1

